test: enable Windows support, un-XFAIL on Linux

compnerd opened the pull request:
Annotate DLL Storage on tests as necessary.  This is needed for Windows
support.  Make some tests pass on Linux as well.  These would be
XFAIL'ed on Windows for the same reason as Linux.  This tightens some
tests, enables some flags, or loosens in other cases.  This allows a
broader set of IRGen tests to run on Linux (and Windows).

Optimize a few tests by collapsing multiple FileCheck invocations or
transferring data through files instead of pipes.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
CC: @aschwaighofer @rjmccall

compnerd commented:
For those following along:

  Expected Passes    : 165
  Unsupported Tests  : 122
  Unexpected Failures: 53

I'm filtering on IRGen first, to make sure that we at least get the IRGen bits tested ASAP.  The rest of the test suite can follow on its heels.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

rjmccall commented on stdlib/public/stubs/MathStubs.cpp:
```diff
@@ -0,0 +1,146 @@
+// runtime library on Linux, even when compiling with clang.
+// This implementation is copied here to avoid a new dependency
+// on compiler-rt on Linux.
+// FIXME: rdar://14883575 Libcompiler_rt omits muloti4
```
This is really unfortunate.  Is avoiding the compiler_rt dependency important for some reason?

compnerd replied:
That was pre-existing.  It is relatively nice to avoid the compiler-rt dependency as it causes a lot of headaches on Linux. libgcc is the more common compiler runtime, and it also includes an unwinder.  On Windows, it also collides with another (static) library - `vcclang.lib`.

compnerd replied:
Note that I didn't rebase the branch, so that commit is not really part of the change here.

compnerd commented:
Current state is:

  Expected Passes    : 207
  Unsupported Tests  : 109
  Unexpected Failures: 30

Most of this is not immediately resolvable, it requires changing the driver and needs the work to port SwiftPrivateStdlibUnittests swift module ported.  I think its time to get this fixed up on Linux and macOS and merged.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd committed:
test: configure the windows test environment

This enables lit to at least start running on Windows.  There is more
work necessary to get the tests actually passing on Windows.

compnerd committed:
test: enable more IRGen tests on Windows

This set of tests also pass on Windows now.  These are mostly invocation
adjustments or renaming/reordering that is needed.

compnerd commented:
Failing Tests (20):
    Swift(windows-x86_64) :: IRGen/associated_type_witness.swift
    Swift(windows-x86_64) :: IRGen/abitypes.swift
    Swift(windows-x86_64) :: IRGen/enum_derived.swift
    Swift(windows-x86_64) :: IRGen/object_type.swift
    Swift(windows-x86_64) :: IRGen/multithread_module.swift
    Swift(windows-x86_64) :: IRGen/dynamic_cast_functions.swift
    Swift(windows-x86_64) :: IRGen/zero_size_types.swift
    Swift(windows-x86_64) :: IRGen/conditional_conformances.swift
    Swift(windows-x86_64) :: IRGen/static_initializer.sil
    Swift(windows-x86_64) :: IRGen/generic_casts.swift
    Swift(windows-x86_64) :: IRGen/existentials_opaque_boxed.sil
    Swift(windows-x86_64) :: IRGen/c_layout.sil
    Swift(windows-x86_64) :: IRGen/weak_import_native.swift
    Swift(windows-x86_64) :: IRGen/meta_meta_type.swift
    Swift(windows-x86_64) :: IRGen/sil_witness_tables.swift
    Swift(windows-x86_64) :: IRGen/vtable_symbol_linkage.swift
    Swift(windows-x86_64) :: IRGen/condfail.sil
    Swift(windows-x86_64) :: IRGen/report_dead_method_call.swift
    Swift(windows-x86_64) :: IRGen/weak_import_native.sil

  Expected Passes    : 217
  Unsupported Tests  : 108
  Unexpected Failures: 19

compnerd commented:
apple/swift#16380 is the last piece from this set.  Going to close this off
